### PR TITLE
[Examples] Add an official audio classification example

### PR DESCRIPTION
**Pull request by anton-l**
This PR adds an example for fine-tuning Wav2Vec2-like models on any labeled audio dataset.

The single-GPU script runs on one V100 and trains on [SUPERB KS](https://huggingface.co/datasets/superb#ks)
The multi-GPU one runs on 4 V100s and trains on [CommonLanguage](https://huggingface.co/datasets/anton-l/common_language) (`datasets` PR: https://github.com/huggingface/datasets/pull/2989)

### CODE REVIEW

File: examples/pytorch/audio-classification/prepare_commonlanguage.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from transformers import logging
+
+
+logging.set_verbosity_info()
+logger = logging.get_logger(__name__)
+
+DATASET_URL = "https://zenodo.org/record/5036977/files/CommonLanguage.tar.gz?download=1"
+
+
+def prepare_commonlanguage(download_directory):
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/prepare_commonlanguage.py:33]: Can't we add the common language to `datasets` https://github.com/huggingface/datasets?

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+    if training_args.do_eval:
+        if "validation" not in ds:
+            raise ValueError("--do_eval requires a validation dataset")
+        if data_args.max_eval_samples is not None:
+            ds["validation"] = (
+                ds["validation"].shuffle(seed=training_args.seed).select(range(data_args.max_eval_samples))
+            )
+        # Set the validation transforms
+        ds["validation"].set_transform(val_transforms, output_all_columns=False)
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/run_audio_classification.py:306]: Do you think loading the audio on the fly is better here?

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    wav, sr = torchaudio.load(path)
+    # convert multi-channel audio to mono
+    wav = wav.mean(0)
+    # standardize sample rate if it varies in the dataset
+    resampler = torchaudio.transforms.Resample(sr, sample_rate)
+    wav = resampler(wav)
+    return wav
+
+
+def random_subsample(wav: np.ndarray, max_length: float, sample_rate: int = 16000):
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/run_audio_classification.py:64]: Does this improve performance you think? Randomly subsampling instead of simple cutting with truncate?

File: examples/pytorch/audio-classification/prepare_commonlanguage.py (hunk 1 above)
Comment (patrickvonplaten) [reply examples/pytorch/audio-classification/prepare_commonlanguage.py:33]: Also this script doesn't seem to be used - why is it here exactly?

File: examples/pytorch/audio-classification/README.md [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+# Audio classification examples
+
+The following examples showcase how to fine-tune `Wav2Vec2` for audio classification using PyTorch.
+
+## Using datasets from 🤗 `datasets`
+
+Here we show how to fine-tune `Wav2Vec2` on the [Keyword Spotting subset](https://huggingface.co/datasets/beans) of the SUPERB dataset.
+
+👀 See the results here: [anton-l/wav2vec2-base-keyword-spotting](https://huggingface.co/anton-l/wav2vec2-base-keyword-spotting).
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/README.md:25]: think this link doesn't exist yet

File: examples/pytorch/audio-classification/run_audio_classification.py (hunk 2 above)
Comment (anton-l) [reply examples/pytorch/audio-classification/run_audio_classification.py]: With `--dataloader_num_workers 4` the GPUs are used close to 100%, so I think it's safe in this case

File: examples/pytorch/audio-classification/run_audio_classification.py (hunk 3 above)
Comment (anton-l) [reply examples/pytorch/audio-classification/run_audio_classification.py:65]: As discussed, in Keyword Spotting there is a `_silence_` class that uses 10min audio clips to sample from, and random sampling is suggested by the SUPERB authors there.

File: examples/pytorch/audio-classification/README.md (hunk 4 above)
Comment (anton-l) [reply examples/pytorch/audio-classification/README.md]: Fixed

**Comment by anton-l** [comment]
Added a multi-gpu example and trained the demo models:
* https://huggingface.co/anton-l/wav2vec2-base-keyword-spotting
* https://huggingface.co/anton-l/wav2vec2-base-langid

Ready for a full review now :)

File: examples/pytorch/audio-classification/README.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    --dataloader_num_workers 4 \
+    --logging_strategy steps \
+    --logging_steps 10 \
+    --evaluation_strategy epoch \
+    --save_strategy epoch \
+    --load_best_model_at_end True \
+    --save_total_limit 3 \
+    --seed 0 \
+    --push_to_hub \
+    --push_to_hub_model_id wav2vec2-base-keyword-spotting
```
Comment (sgugger) [inline examples/pytorch/audio-classification/README.md:57]: The `push_to_hub_model_id` argument is deprecated, so you should name the `output_dir` `wav2vec2-base-keyword-spotting` and just remove that line.

File: examples/pytorch/audio-classification/README.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    --dataloader_num_workers 8 \
+    --logging_strategy steps \
+    --logging_steps 10 \
+    --evaluation_strategy epoch \
+    --save_strategy epoch \
+    --load_best_model_at_end True \
+    --save_total_limit 3 \
+    --seed 0 \
+    --push_to_hub \
+    --push_to_hub_model_id wav2vec2-base-langid
```
Comment (sgugger) [inline examples/pytorch/audio-classification/README.md:96]: Same here

File: examples/pytorch/audio-classification/README.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+## Sharing your model on 🤗 Hub
+
+0. If you haven't already, [sign up](https://huggingface.co/join) for a 🤗 account
+
+1. Make sure you have `git-lfs` installed and git set up.
+
+```bash
+$ apt install git-lfs
+$ git config --global user.email "you@example.com"
+$ git config --global user.name "Your Name"
```
Comment (sgugger) [inline examples/pytorch/audio-classification/README.md:135]: This shouldn't be necessary anymore.

File: examples/pytorch/audio-classification/README.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+$ huggingface-cli login
+# ...follow the prompts
+```
+
+3. When running the script, pass the following arguments:
+
+```bash
+python run_audio_classification.py \
+    --push_to_hub \
+    --push_to_hub_model_id <name-your-model> \
```
Comment (sgugger) [inline examples/pytorch/audio-classification/README.md:150]: The right argument is `hub_model_id` now, and it should be the full model name (e.g. username/model_id)

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+        )
+
+    if data_args.label_column_name not in ds["train"].column_names:
+        raise ValueError(
+            f"--label_column_name {data_args.label_column_name} not found in dataset '{data_args.dataset_name}'. "
+            "Make sure to set `--label_column_name` to the correct text column - one of "
+            f"{', '.join(ds['train'].column_names)}."
+        )
+
+    feature_extractor = Wav2Vec2FeatureExtractor.from_pretrained(
```
Comment (sgugger) [inline examples/pytorch/audio-classification/run_audio_classification.py:248]: `AutoFeatureExtractor` should be used here since it's `AutoModelForSpeech` later on.

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+        id2label[str(i)] = label
+
+    # Load the accuracy metric from the datasets package
+    metric = datasets.load_metric("accuracy")
+
+    # Define our compute_metrics function. It takes an `EvalPrediction` object (a namedtuple with a
+    # predictions and label_ids field) and has to return a dictionary string to float.
+    def compute_metrics(p):
+        """Computes accuracy on a batch of predictions"""
+        return metric.compute(predictions=np.argmax(p.predictions, axis=1), references=p.label_ids)
```
Comment (sgugger) [inline examples/pytorch/audio-classification/run_audio_classification.py:294]:
```suggestion
        predictions = np.argmax(p.predictions, axis=1)
        return metric.compute(predictions=predictions, references=p.label_ids)
```
Let's split the argmax in another line for easier reading.

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+        elif last_checkpoint is not None and training_args.resume_from_checkpoint is None:
+            logger.info(
+                f"Checkpoint detected, resuming training at {last_checkpoint}. To avoid this behavior, change "
+                "the `--output_dir` or add `--overwrite_output_dir` to train from scratch."
+            )
+
+    # Initialize our dataset and prepare it for the audio classification task.
+    ds = DatasetDict()
+    ds["train"] = load_dataset(data_args.dataset_name, data_args.dataset_config_name, split=data_args.train_split_name)
+    ds["eval"] = load_dataset(data_args.dataset_name, data_args.dataset_config_name, split=data_args.eval_split_name)
```
Comment (sgugger) [inline examples/pytorch/audio-classification/run_audio_classification.py:232]: Let's not use something as short as `ds`, as this is very much against the usual Hugging Face style for a variable that is used in the whole file.
Other scripts use `raw_datasets` then `tokenized_datasets` for instance.

**Review by sgugger — APPROVED** [review]
Thanks a lot for adding this! I have left a few comments.

Could you rebase your branch on master? You don't seem to have the last fix on the test fetcher utils for the examples, so your new example test is actually not run.

**Review by patil-suraj — APPROVED** [review]
LGTM! Thanks for adding this example

File: examples/pytorch/audio-classification/README.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+The following examples showcase how to fine-tune `Wav2Vec2` for audio classification using PyTorch.
+
+Speech recognition models that have been pretrained in unsupervised fashion on audio data alone, *e.g.* [Wav2Vec2](https://huggingface.co/transformers/master/model_doc/wav2vec2.html), [HuBERT](https://huggingface.co/transformers/master/model_doc/hubert.html), [XLSR-Wav2Vec2](https://huggingface.co/transformers/master/model_doc/xlsr_wav2vec2.html), have shown to require only 
+very little annotated data to yield good performance on speech classification datasets.
+
+## Single-GPU 
+
+Here we show how to fine-tune [wav2vec2-base](https://huggingface.co/facebook/wav2vec2-base) on the 🗣️ [Keyword Spotting subset](https://huggingface.co/datasets/superb#ks) of the SUPERB dataset.
+
+👀 See the results here: [anton-l/wav2vec2-base-keyword-spotting](https://huggingface.co/anton-l/wav2vec2-base-keyword-spotting)
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/README.md:28]: (nit) I would maybe move this line below the bash command. It's a bit more natural for me to see how to do it and then the result of how it should look like

File: examples/pytorch/audio-classification/README.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+    --push_to_hub
+```
+
+On a single V100 GPU (16GB), this script should run in ~10 minutes and yield accuracy of **98.4%**.
+
+## Multi-GPU 
+
+Here we show how to fine-tune [wav2vec2-base](https://huggingface.co/facebook/wav2vec2-base) for 🌎 **Language Identification** on the [CommonLanguage dataset](https://huggingface.co/datasets/beans).
+
+👀 See the results here: [anton-l/wav2vec2-base-langid](https://huggingface.co/anton-l/wav2vec2-base-langid)
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/README.md:65]: (nit) same

File: examples/pytorch/audio-classification/README.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    --save_strategy epoch \
+    --load_best_model_at_end True \
+    --save_total_limit 3 \
+    --seed 0 \
+    --push_to_hub
+```
+
+On 4 V100 GPUs (16GB), this script should run in ~1 hour and yield accuracy of **77.51%**.
+
+## Sharing your model on 🤗 Hub
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/README.md:99]: Cool!

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 15]
```diff
@@ -0,0 +1,379 @@
...
+class ModelArguments:
+    """
+    Arguments pertaining to which model/config/tokenizer we are going to fine-tune from.
+    """
+
+    model_name_or_path: str = field(
+        default="facebook/wav2vec2-base",
+        metadata={"help": "Path to pretrained model or model identifier from huggingface.co/models"},
+    )
+    model_type: Optional[str] = field(
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/run_audio_classification.py:143]: I don't this this param is used no?

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+    if data_args.label_column_name not in raw_datasets["train"].column_names:
+        raise ValueError(
+            f"--label_column_name {data_args.label_column_name} not found in dataset '{data_args.dataset_name}'. "
+            "Make sure to set `--label_column_name` to the correct text column - one of "
+            f"{', '.join(raw_datasets['train'].column_names)}."
+        )
+
+    feature_extractor = AutoFeatureExtractor.from_pretrained(
+        model_args.feature_extractor_name or model_args.model_name_or_path,
+        return_attention_mask=True,
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/run_audio_classification.py:246]: We should change this here. If we change the fetaure extractor config by passing `return_attention_mask=True` then we also have to save the feature extractor with the config having `return_attention_mask=True`. Otherwise people will train: https://huggingface.co/facebook/wav2vec2-base/blob/main/preprocessor_config.json (which has `return_attention_mask==False`), not save it and use the model with the default feature extractor which will lead to incorrect results. 

The problem with wav2vec2-base is that batched inference != non-batched inference by model design. In my experiments of wav2vec2-base on ASR, I found that not using the attention mask led to better results. 
Think we have to options here:

1) Not pass `return_attention_mask=True` - we would probably have to quickly re-run the experiments to see if the results are good with this as well

2) Save the feature extractor alongside the model

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+        for f in batch[data_args.audio_column_name]:
+            wav = load_audio(f, sample_rate=feature_extractor.sampling_rate)
+            output_batch["input_values"].append(wav)
+        output_batch["labels"] = [label for label in batch[data_args.label_column_name]]
+
+        return output_batch
+
+    # Prepare label mappings.
+    # We'll include these in the model's config to get human readable labels in the Inference API.
+    labels = raw_datasets["train"].features[data_args.label_column_name].names
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/run_audio_classification.py:277]: very nice!

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 18]
```diff
@@ -0,0 +1,379 @@
...
+    for i, label in enumerate(labels):
+        label2id[label] = str(i)
+        id2label[str(i)] = label
+
+    # Load the accuracy metric from the datasets package
+    metric = datasets.load_metric("accuracy")
+
+    # Define our compute_metrics function. It takes an `EvalPrediction` object (a namedtuple with a
+    # predictions and label_ids field) and has to return a dictionary string to float.
+    def compute_metrics(p):
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/run_audio_classification.py:288]: not a huge fan of single not-annoted function arguments here...could you give the variable a nicer name? :-)

File: examples/pytorch/audio-classification/run_audio_classification.py [hunk 19]
```diff
@@ -0,0 +1,379 @@
...
+        """Computes accuracy on a batch of predictions"""
+        predictions = np.argmax(p.predictions, axis=1)
+        return metric.compute(predictions=predictions, references=p.label_ids)
+
+    config = AutoConfig.from_pretrained(
+        model_args.config_name or model_args.model_name_or_path,
+        num_labels=len(labels),
+        label2id=label2id,
+        id2label=id2label,
+        finetuning_task="audio-classification",
```
Comment (patrickvonplaten) [inline examples/pytorch/audio-classification/run_audio_classification.py:298]: nice!

**Review by patrickvonplaten — COMMENTED** [review]
Very nice addition! -

Before merging I think we should make sure though that the passed config variable `return_attention_mask` is either saved or removed :-) Then we're good for merge IMO

File: examples/pytorch/audio-classification/run_audio_classification.py (hunk 16 above)
Comment (anton-l) [reply examples/pytorch/audio-classification/run_audio_classification.py]: In this case the feature extractor is passed as `tokenizer=feature_extractor` to `Trainer`, so it's saved alongside the model: https://huggingface.co/anton-l/wav2vec2-base-langid/blob/main/preprocessor_config.json

I've added an `--attention_mask` flag and re-done the CommonLanguage experiment with `--attention_mask=False`, and the result is indeed slightly better :)

Comment (patrickvonplaten) [reply examples/pytorch/audio-classification/run_audio_classification.py]: Sorry, you're 100% right - the `Trainer` actually saves the `feature_extractor`! Ok in this case - all good for me :-)

**Comment by patrickvonplaten** [comment]
Very nice addition!
